streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -623,63 +642,75 @@ void runOnce() {
-                        break;
-                    }
+                final int processed = taskManager.process(numIterations, now);
+                if (processed > 0) {
+                    // It makes no difference to the outcome of these metrics when we record "0",
+                    // so we can just avoid the method call when we didn't process anything.
+                    processRateSensor.record(processed, now);
+
+                    // This metric is scaled to represent the _average_ processing time of _each_
+                    // task. Note, it's hard to interpret this as defined, but we would need a KIP
+                    // to change it to simply report the overall time spent processing all tasks.
+                    final long processLatency = advanceNowAndComputeLatency();
+                    processLatencySensor.record(processLatency / (double) processed, now);
+                    totalProcessLatency += processLatency;
+                }
+
+                final int punctuated = taskManager.punctuate();
+                if (punctuated > 0) {
+                    final long punctuateLatency = advanceNowAndComputeLatency();
+                    punctuateSensor.record(punctuateLatency / (double) punctuated, now);
+                    totalPunctuateLatency += punctuateLatency;
                 }
 
-                timeSinceLastPoll = Math.max(now - lastPollMs, 0);
+                final int committed = maybeCommit();
+                if (committed > 0) {
+                    final long commitLatency = advanceNowAndComputeLatency();
+                    commitSensor.record(commitLatency / (double) committed, now);
 
-                if (maybePunctuate() || maybeCommit()) {
-                    numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
-                } else if (timeSinceLastPoll > maxPollTimeMs / 2) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Committed all active tasks {} and standby tasks {} in {}ms",
+                            taskManager.activeTaskIds(), taskManager.standbyTaskIds(), commitLatency);
+                    }
+
+                    totalCommitLatency += commitLatency;
+                }
+
+                if (processed == 0) {
+                    // if there is no records to be processed, exit after punctuate / commit
+                    break;
+                } else if (Math.max(now - lastPollMs, 0) > maxPollTimeMs / 2) {
                     numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
                     break;
-                } else if (processed > 0) {
+                } else if (punctuated > 0 || committed > 0) {
+                    numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
+                } else {
                     numIterations++;
                 }
-            } while (processed > 0);
-
-            maybeCommit();
```

Comment:
This is a minor fix: previously we need to call this because we process standbys AFTER processing actives and hence may be we need to commit again afterwards, but not we process standbys first, then active, so this can be removed.